src/sources/soundsourceffmpeg.cpp

```diff
@@ -304,672 +416,843 @@ SoundSourceFFmpeg::~SoundSourceFFmpeg() {
-                        m_SCache.append(l_SObj);
-                        l_SObj->startFrame = m_lCacheFramePos;
-                        l_SObj->length = l_iRet;
-                        m_lCacheFramePos += AUDIOSOURCEFFMPEG_BYTEOFFSET_TO_MIXXXFRAME(l_iRet);
-
-                        // Ogg/Opus have packages pos that have many
-                        // audio frames so seek next unique pos..
-                        if (l_SPacket.pos != l_lLastPacketPos) {
-                            l_lLastPacketPos = l_SPacket.pos;
-                            m_lStoreCount++;
-                        }
-
-                        // If we are over last storepos and we have read more than jump point need is
-                        // and pos is unique we store it to memory
-                        //
-                        // So why number 32? Sorry I just made that up from the hat there is no
-                        // math behind it. Number 32 it's not too big nor jumps points are not
-                        // too close each other. Mainly it's ugly compromise with MP3,MP4,OGG and WMA
-                        // different codec frame sizes
-                        if (m_lStoreCount == 32) {
-                            struct ffmpegLocationObject* l_STestObj = nullptr;
-
-                            if (m_SJumpPoints.size() > 0) {
-                                l_STestObj = m_SJumpPoints.last();
-                            }
-                            // Check whether we have this jumppoint stored already or not
-                            // We should have jumppoints below that pos
-                            if (l_STestObj == nullptr || l_STestObj->pos < l_SPacket.pos) {
-                                struct ffmpegLocationObject* l_SJmp = (struct ffmpegLocationObject*)malloc(
-                                        sizeof(struct ffmpegLocationObject));
-                                m_lLastStoredPos = m_lCacheFramePos;
-                                l_SJmp->startFrame = m_lCacheFramePos;
-                                l_SJmp->pos = l_SPacket.pos;
-                                l_SJmp->pts = l_SPacket.pts;
-                                m_SJumpPoints.append(l_SJmp);
-                            }
-                            m_lStoreCount = 0;
-                        }
-
-                        if (offset < 0 || offset <= m_lCacheFramePos) {
-                            l_iCount--;
-                        }
-                    } else {
-                        free(l_SObj);
-                        l_SObj = nullptr;
-                        kLogger.debug() << "readFramesToCache: General error in audio decode:" << l_iRet;
-                    }
-                }
-
-#if (LIBAVCODEC_HAS_AV_PACKET_UNREF)
-                av_packet_unref(&l_SPacket);
-#else
-                av_free_packet(&l_SPacket);
+ReadableSampleFrames SoundSourceFFmpeg::readSampleFramesClamped(
+        WritableSampleFrames writableSampleFrames) {
```

Comment:
We use the term *clamped* to indicate that the given index range is guaranteed to be within the min/max bounds of the stream. These checks and adjustments are done by the base class instead of doing this again and again in each derived class -> _Template Method_ pattern.